File: crates/driftguard-core/src/validation.rs
//! Phase 4 validation pipeline: run the eval suite, judge it, and label ground
//! truth — the steps that turn fixtures into a measurable precision/recall result.
//!
//! Flow per edit (before = base v1, after = the edit's version):
//!   1. `run-evals` each version once (generate actual_output for every case).
//!   2. Judge Question A on every run (did it satisfy expected_behavior?).
//!   3. Judge Question B per case (did behavior change before→after?) → BehaviorDiff.
//!   4. Run similarity selection on the after-version (Phase 3).
//!   5. Label SelectionRecord.was_actually_affected = the case's behavior-changed
//!      verdict (the chosen ground truth).
//!
//! Calls are sequential (your decision): simplest, no rate-limit handling.

use serde::Serialize;
use uuid::Uuid;

use crate::db::Pool;
use crate::embed::Embedder;
use crate::error::CoreError;
use crate::evals::list_eval_cases;
use crate::fixtures::Fixtures;
use crate::llm::Llm;
use crate::models::{EvalCase, Prompt};
use crate::prompts::{get_prompt_by_name, resolve_version_ref};
use crate::selection;

#[derive(Debug, Default, Serialize)]
pub struct ValidationSummary {
    pub prompts: usize,
    pub edits: usize,
    pub versions_evaluated: usize,
    pub eval_runs: usize,
    pub behavior_diffs: usize,
}

/// Generate actual_output for every eval case against one version, overwriting
/// any prior runs for that version (idempotent re-runs — your decision).
pub async fn run_evals<L: Llm>(
    pool: &Pool,
    llm: &L,
    prompt_name: &str,
    version_ref: &str,
) -> Result<usize, CoreError> {
    let prompt = get_prompt_by_name(pool, prompt_name)
        .await?
        .ok_or_else(|| CoreError::PromptNotFound(prompt_name.to_string()))?;
    let version = resolve_version_ref(pool, &prompt, version_ref).await?;
    let cases = list_eval_cases(pool, prompt.id).await?;

    sqlx::query!(
        "DELETE FROM eval_runs WHERE prompt_version_id = $1",
        version.id
    )
    .execute(pool)
    .await?;

    for case in &cases {
        let output = llm.generate(&version.content, &case.input).await?;
        sqlx::query!(
            "INSERT INTO eval_runs (prompt_version_id, eval_case_id, actual_output)
             VALUES ($1, $2, $3)",
            version.id,
            case.id,
            output
        )
        .execute(pool)
        .await?;
    }
    Ok(cases.len())
}

/// Run the entire fixture matrix end to end and return a summary.
pub async fn validate_fixtures<L: Llm, E: Embedder>(
    pool: &Pool,
    llm: &L,
    embedder: &E,
    fixtures: &Fixtures,
    threshold: f64,
) -> Result<ValidationSummary, CoreError> {
    let mut summary = ValidationSummary::default();

    for fp in &fixtures.prompts {
        let prompt = get_prompt_by_name(pool, &fp.name).await?.ok_or_else(|| {
            CoreError::InvalidState(format!(
                "prompt \"{}\" is not loaded; run `fixtures load` first",
                fp.name
            ))
        })?;
        let cases = list_eval_cases(pool, prompt.id).await?;
        summary.prompts += 1;

        // Base version (v1): generate + judge A once.
        run_evals(pool, llm, &fp.name, "v1").await?;
        judge_pass_for_version(pool, llm, &prompt, &cases, "v1").await?;
        summary.versions_evaluated += 1;
        summary.eval_runs += cases.len();

        // Each edit branched off v1 → its version is v{index+2}.
        for (index, _edit) in fp.edits.iter().enumerate() {
            let after_ref = format!("v{}", index + 2);

            run_evals(pool, llm, &fp.name, &after_ref).await?;
            judge_pass_for_version(pool, llm, &prompt, &cases, &after_ref).await?;
            behavior_diffs_for_edit(pool, llm, &prompt, &cases, "v1", &after_ref).await?;

            // Phase 3 selection on the changed version, then label ground truth.
            selection::select_evals(pool, embedder, &fp.name, &after_ref, threshold).await?;
            label_ground_truth(pool, &prompt, &cases, "v1", &after_ref).await?;

            summary.edits += 1;
            summary.versions_evaluated += 1;
            summary.eval_runs += cases.len();
            summary.behavior_diffs += cases.len();
        }
    }

    Ok(summary)
}

/// Judge Question A for every eval run of one version.
async fn judge_pass_for_version<L: Llm>(
    pool: &Pool,
    llm: &L,
    prompt: &Prompt,
    cases: &[EvalCase],
    version_ref: &str,
) -> Result<(), CoreError> {
    let version = resolve_version_ref(pool, prompt, version_ref).await?;
    for case in cases {
        let output = fetch_output(pool, version.id, case.id).await?;
        let verdict = llm
            .judge_pass(&case.expected_behavior, &case.input, &output)
            .await?;
        sqlx::query!(
            "UPDATE eval_runs SET judge_passed = $1, judge_justification = $2
             WHERE prompt_version_id = $3 AND eval_case_id = $4",
            verdict.passed,
            verdict.justification,
            version.id,
            case.id
        )
        .execute(pool)
        .await?;
    }
    Ok(())
}

/// Judge Question B for an edit, writing one BehaviorDiff per case (overwriting
/// any prior diff for the same before/after/case triple).
async fn behavior_diffs_for_edit<L: Llm>(
    pool: &Pool,
    llm: &L,
    prompt: &Prompt,
    cases: &[EvalCase],
    before_ref: &str,
    after_ref: &str,
) -> Result<(), CoreError> {
    let before = resolve_version_ref(pool, prompt, before_ref).await?;
    let after = resolve_version_ref(pool, prompt, after_ref).await?;

    for case in cases {
        let before_output = fetch_output(pool, before.id, case.id).await?;
        let after_output = fetch_output(pool, after.id, case.id).await?;
        let verdict = llm
            .judge_behavior(
                &case.input,
                &case.expected_behavior,
                &before_output,
                &after_output,
            )
            .await?;

        sqlx::query!(
            "DELETE FROM behavior_diffs
             WHERE eval_case_id = $1 AND prompt_version_before_id = $2
               AND prompt_version_after_id = $3",
            case.id,
            before.id,
            after.id
        )
        .execute(pool)
        .await?;
        sqlx::query!(
            "INSERT INTO behavior_diffs
                 (eval_case_id, prompt_version_before_id, prompt_version_after_id,
                  judge_behavior_changed, judge_justification)
             VALUES ($1, $2, $3, $4, $5)",
            case.id,
            before.id,
            after.id,
            verdict.behavior_changed,
            verdict.justification
        )
        .execute(pool)
        .await?;
    }
    Ok(())
}

/// Copy each case's behavior-changed verdict onto its SelectionRecord as ground
/// truth (was_actually_affected).
async fn label_ground_truth(
    pool: &Pool,
    prompt: &Prompt,
    cases: &[EvalCase],
    before_ref: &str,
    after_ref: &str,
) -> Result<(), CoreError> {
    let before = resolve_version_ref(pool, prompt, before_ref).await?;
    let after = resolve_version_ref(pool, prompt, after_ref).await?;

    for case in cases {
        let changed = sqlx::query_scalar!(
            "SELECT judge_behavior_changed FROM behavior_diffs
             WHERE eval_case_id = $1 AND prompt_version_before_id = $2
               AND prompt_version_after_id = $3",
            case.id,
            before.id,
            after.id
        )
        .fetch_optional(pool)
        .await?
        .flatten();

        sqlx::query!(
            "UPDATE selection_records SET was_actually_affected = $1
             WHERE prompt_version_id = $2 AND eval_case_id = $3",
            changed,
            after.id,
            case.id
        )
        .execute(pool)
        .await?;
    }
    Ok(())
}

async fn fetch_output(pool: &Pool, version_id: Uuid, case_id: Uuid) -> Result<String, CoreError> {
    sqlx::query_scalar!(
        "SELECT actual_output FROM eval_runs
         WHERE prompt_version_id = $1 AND eval_case_id = $2",
        version_id,
        case_id
    )
    .fetch_optional(pool)
    .await?
    .ok_or_else(|| CoreError::InvalidState("missing eval run (run-evals first)".to_string()))
}

// ---- judge-validate support (human spot-check of judge verdicts) ----

#[derive(Debug, Serialize)]
pub struct DiffSample {
    pub behavior_diff_id: Uuid,
    pub input: String,
    pub expected_behavior: String,
    pub before_output: String,
    pub after_output: String,
    pub judge_behavior_changed: Option<bool>,
    pub judge_justification: Option<String>,
}

/// Pull a random sample of judge verdicts that haven't been human-reviewed yet.
pub async fn sample_unvalidated_diffs(
    pool: &Pool,
    sample_size: i64,
) -> Result<Vec<DiffSample>, CoreError> {
    let rows = sqlx::query!(
        r#"SELECT bd.id AS "behavior_diff_id!",
                  ec.input AS "input!",
                  ec.expected_behavior AS "expected_behavior!",
                  before_run.actual_output AS "before_output!",
                  after_run.actual_output AS "after_output!",
                  bd.judge_behavior_changed,
                  bd.judge_justification
           FROM behavior_diffs bd
           JOIN eval_cases ec ON ec.id = bd.eval_case_id
           JOIN eval_runs before_run
                ON before_run.prompt_version_id = bd.prompt_version_before_id
               AND before_run.eval_case_id = bd.eval_case_id
           JOIN eval_runs after_run
                ON after_run.prompt_version_id = bd.prompt_version_after_id
               AND after_run.eval_case_id = bd.eval_case_id
           WHERE bd.human_agreed IS NULL
           ORDER BY random()
           LIMIT $1"#,
        sample_size
    )
    .fetch_all(pool)
    .await?;

    Ok(rows
        .into_iter()
        .map(|r| DiffSample {
            behavior_diff_id: r.behavior_diff_id,
            input: r.input,
            expected_behavior: r.expected_behavior,
            before_output: r.before_output,
            after_output: r.after_output,
            judge_behavior_changed: r.judge_behavior_changed,
            judge_justification: r.judge_justification,
        })
        .collect())
}

/// Record a human's agree/disagree with a judge verdict.
pub async fn set_human_agreed(
    pool: &Pool,
    behavior_diff_id: Uuid,
    agreed: bool,
) -> Result<(), CoreError> {
    sqlx::query!(
        "UPDATE behavior_diffs SET human_agreed = $1 WHERE id = $2",
        agreed,
        behavior_diff_id
    )
    .execute(pool)
    .await?;
    Ok(())
}

/// (agreed, total reviewed) across all human-reviewed verdicts.
pub async fn judge_agreement(pool: &Pool) -> Result<(i64, i64), CoreError> {
    let row = sqlx::query!(
        r#"SELECT
             count(*) FILTER (WHERE human_agreed) AS "agreed!",
             count(*) FILTER (WHERE human_agreed IS NOT NULL) AS "total!"
           FROM behavior_diffs"#
    )
    .fetch_one(pool)
    .await?;
    Ok((row.agreed, row.total))
}

File: internal/core/validation.go
package core

// Validation pipeline (port of validation.rs): run the eval suite, judge it,
// and label ground truth — the steps that turn fixtures into a measurable
// precision/recall result.
//
// Flow per edit (before = base v1, after = the edit's version):
//  1. run-evals each version once (generate actual_output for every case).
//  2. Judge Question A on every run (did it satisfy expected_behavior?).
//  3. Judge Question B per case (did behavior change before→after?).
//  4. Run similarity selection on the after-version.
//  5. Label selection_records.was_actually_affected with the behavior verdict.
//
// LLM calls are bounded-concurrent via errgroup (the Rust version used
// buffer_unordered over the sidecar; here it's plain goroutines).

import (
	"context"
	"fmt"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
	"golang.org/x/sync/errgroup"
)

type ValidationSummary struct {
	Prompts           int `json:"prompts"`
	Edits             int `json:"edits"`
	VersionsEvaluated int `json:"versions_evaluated"`
	EvalRuns          int `json:"eval_runs"`
	BehaviorDiffs     int `json:"behavior_diffs"`
}

// RunEvals generates actual_output for every eval case against one version,
// overwriting any prior runs for that version. concurrency calls in flight.
func RunEvals(ctx context.Context, pool *pgxpool.Pool, llm Llm, promptName, versionRef string, concurrency int) (int, error) {
	prompt, err := GetPromptByName(ctx, pool, promptName)
	if err != nil {
		return 0, err
	}
	if prompt == nil {
		return 0, PromptNotFoundError{Name: promptName}
	}
	version, err := ResolveVersionRef(ctx, pool, prompt, versionRef)
	if err != nil {
		return 0, err
	}
	cases, err := ListEvalCases(ctx, pool, prompt.ID)
	if err != nil {
		return 0, err
	}

	if _, err := pool.Exec(ctx,
		`DELETE FROM eval_runs WHERE prompt_version_id = $1`, version.ID); err != nil {
		return 0, err
	}

	g, gctx := errgroup.WithContext(ctx)
	g.SetLimit(maxInt(concurrency, 1))
	for _, c := range cases {
		g.Go(func() error {
			output, err := llm.Generate(gctx, version.Content, c.Input)
			if err != nil {
				return err
			}
			_, err = pool.Exec(gctx,
				`INSERT INTO eval_runs (prompt_version_id, eval_case_id, actual_output)
				 VALUES ($1, $2, $3)`, version.ID, c.ID, output)
			return err
		})
	}
	if err := g.Wait(); err != nil {
		return 0, err
	}
	return len(cases), nil
}

// ValidateFixtures runs the entire fixture matrix end to end.
func ValidateFixtures(ctx context.Context, pool *pgxpool.Pool, llm Llm, embedder Embedder, fixtures *Fixtures, threshold float64, concurrency int) (*ValidationSummary, error) {
	summary := &ValidationSummary{}

	for _, fp := range fixtures.Prompts {
		prompt, err := GetPromptByName(ctx, pool, fp.Name)
		if err != nil {
			return nil, err
		}
		if prompt == nil {
			return nil, fmt.Errorf("prompt %q is not loaded; run `fixtures load` first", fp.Name)
		}
		cases, err := ListEvalCases(ctx, pool, prompt.ID)
		if err != nil {
			return nil, err
		}
		summary.Prompts++

		// Base version (v1): generate + judge A once.
		if _, err := RunEvals(ctx, pool, llm, fp.Name, "v1", concurrency); err != nil {
			return nil, err
		}
		if err := judgePassForVersion(ctx, pool, llm, prompt, cases, "v1", concurrency); err != nil {
			return nil, err
		}
		summary.VersionsEvaluated++
		summary.EvalRuns += len(cases)

		// Each edit branched off v1 → its version is v{index+2}.
		for index := range fp.Edits {
			afterRef := fmt.Sprintf("v%d", index+2)

			if _, err := RunEvals(ctx, pool, llm, fp.Name, afterRef, concurrency); err != nil {
				return nil, err
			}
			if err := judgePassForVersion(ctx, pool, llm, prompt, cases, afterRef, concurrency); err != nil {
				return nil, err
			}
			if err := behaviorDiffsForEdit(ctx, pool, llm, prompt, cases, "v1", afterRef, concurrency); err != nil {
				return nil, err
			}

			// Selection on the changed version, then label ground truth.
			if _, err := SelectEvals(ctx, pool, embedder, fp.Name, afterRef, threshold); err != nil {
				return nil, err
			}
			if err := labelGroundTruth(ctx, pool, prompt, cases, "v1", afterRef); err != nil {
				return nil, err
			}

			summary.Edits++
			summary.VersionsEvaluated++
			summary.EvalRuns += len(cases)
			summary.BehaviorDiffs += len(cases)
		}
	}

	return summary, nil
}

// judgePassForVersion judges Question A for every eval run of one version.
func judgePassForVersion(ctx context.Context, pool *pgxpool.Pool, llm Llm, prompt *Prompt, cases []EvalCase, versionRef string, concurrency int) error {
	version, err := ResolveVersionRef(ctx, pool, prompt, versionRef)
	if err != nil {
		return err
	}
	g, gctx := errgroup.WithContext(ctx)
	g.SetLimit(maxInt(concurrency, 1))
	for _, c := range cases {
		g.Go(func() error {
			output, err := fetchOutput(gctx, pool, version.ID, c.ID)
			if err != nil {
				return err
			}
			verdict, err := llm.JudgePass(gctx, c.ExpectedBehavior, c.Input, output)
			if err != nil {
				return err
			}
			_, err = pool.Exec(gctx,
				`UPDATE eval_runs SET judge_passed = $1, judge_justification = $2
				 WHERE prompt_version_id = $3 AND eval_case_id = $4`,
				verdict.Passed, verdict.Justification, version.ID, c.ID)
			return err
		})
	}
	return g.Wait()
}

// behaviorDiffsForEdit judges Question B for an edit, writing one behavior
// diff per case (overwriting any prior diff for the same triple).
func behaviorDiffsForEdit(ctx context.Context, pool *pgxpool.Pool, llm Llm, prompt *Prompt, cases []EvalCase, beforeRef, afterRef string, concurrency int) error {
	before, err := ResolveVersionRef(ctx, pool, prompt, beforeRef)
	if err != nil {
		return err
	}
	after, err := ResolveVersionRef(ctx, pool, prompt, afterRef)
	if err != nil {
		return err
	}

	g, gctx := errgroup.WithContext(ctx)
	g.SetLimit(maxInt(concurrency, 1))
	for _, c := range cases {
		g.Go(func() error {
			beforeOutput, err := fetchOutput(gctx, pool, before.ID, c.ID)
			if err != nil {
				return err
			}
			afterOutput, err := fetchOutput(gctx, pool, after.ID, c.ID)
			if err != nil {
				return err
			}
			verdict, err := llm.JudgeBehavior(gctx, c.Input, c.ExpectedBehavior, beforeOutput, afterOutput)
			if err != nil {
				return err
			}

			if _, err := pool.Exec(gctx,
				`DELETE FROM behavior_diffs
				 WHERE eval_case_id = $1 AND prompt_version_before_id = $2
				   AND prompt_version_after_id = $3`, c.ID, before.ID, after.ID); err != nil {
				return err
			}
			_, err = pool.Exec(gctx,
				`INSERT INTO behavior_diffs
				     (eval_case_id, prompt_version_before_id, prompt_version_after_id,
				      judge_behavior_changed, judge_justification)
				 VALUES ($1, $2, $3, $4, $5)`,
				c.ID, before.ID, after.ID, verdict.BehaviorChanged, verdict.Justification)
			return err
		})
	}
	return g.Wait()
}

// labelGroundTruth copies each case's behavior-changed verdict onto its
// selection record as ground truth (was_actually_affected).
func labelGroundTruth(ctx context.Context, pool *pgxpool.Pool, prompt *Prompt, cases []EvalCase, beforeRef, afterRef string) error {
	before, err := ResolveVersionRef(ctx, pool, prompt, beforeRef)
	if err != nil {
		return err
	}
	after, err := ResolveVersionRef(ctx, pool, prompt, afterRef)
	if err != nil {
		return err
	}

	for _, c := range cases {
		var changed *bool
		err := pool.QueryRow(ctx,
			`SELECT judge_behavior_changed FROM behavior_diffs
			 WHERE eval_case_id = $1 AND prompt_version_before_id = $2
			   AND prompt_version_after_id = $3`, c.ID, before.ID, after.ID).Scan(&changed)
		if err != nil && err.Error() != "no rows in result set" {
			return err
		}

		if _, err := pool.Exec(ctx,
			`UPDATE selection_records SET was_actually_affected = $1
			 WHERE prompt_version_id = $2 AND eval_case_id = $3`,
			changed, after.ID, c.ID); err != nil {
			return err
		}
	}
	return nil
}

func fetchOutput(ctx context.Context, pool *pgxpool.Pool, versionID, caseID uuid.UUID) (string, error) {
	var output *string
	err := pool.QueryRow(ctx,
		`SELECT actual_output FROM eval_runs
		 WHERE prompt_version_id = $1 AND eval_case_id = $2`, versionID, caseID).Scan(&output)
	if err != nil || output == nil {
		return "", fmt.Errorf("missing eval run (run-evals first)")
	}
	return *output, nil
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// ---- hand-labeling support (independent human ground truth) ----

// DiffSample is one behavior-diff presented for hand-labeling: the input, the
// two actual outputs, the judge's verdict (shown for reference, NOT to anchor
// the labeler), and the current human label if any.
type DiffSample struct {
	BehaviorDiffID       uuid.UUID `json:"behavior_diff_id"`
	Input                string    `json:"input"`
	ExpectedBehavior     string    `json:"expected_behavior"`
	BeforeOutput         string    `json:"before_output"`
	AfterOutput          string    `json:"after_output"`
	JudgeBehaviorChanged *bool     `json:"judge_behavior_changed"`
	JudgeJustification   *string   `json:"judge_justification"`
	HumanBehaviorChanged *bool     `json:"human_behavior_changed"`
}

// SampleUnlabeledDiffs pulls a random sample of behavior-diffs that don't yet
// have a human label. sampleSize <= 0 returns all of them.
func SampleUnlabeledDiffs(ctx context.Context, pool *pgxpool.Pool, sampleSize int64) ([]DiffSample, error) {
	limit := "ALL"
	args := []any{}
	if sampleSize > 0 {
		limit = "$1"
		args = append(args, sampleSize)
	}
	rows, err := pool.Query(ctx, `
		SELECT bd.id, ec.input, ec.expected_behavior,
		       before_run.actual_output, after_run.actual_output,
		       bd.judge_behavior_changed, bd.judge_justification, bd.human_behavior_changed
		FROM behavior_diffs bd
		JOIN eval_cases ec ON ec.id = bd.eval_case_id
		JOIN eval_runs before_run
		     ON before_run.prompt_version_id = bd.prompt_version_before_id
		    AND before_run.eval_case_id = bd.eval_case_id
		JOIN eval_runs after_run
		     ON after_run.prompt_version_id = bd.prompt_version_after_id
		    AND after_run.eval_case_id = bd.eval_case_id
		WHERE bd.human_behavior_changed IS NULL
		ORDER BY random()
		LIMIT `+limit, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	samples := []DiffSample{}
	for rows.Next() {
		var s DiffSample
		if err := rows.Scan(&s.BehaviorDiffID, &s.Input, &s.ExpectedBehavior,
			&s.BeforeOutput, &s.AfterOutput, &s.JudgeBehaviorChanged, &s.JudgeJustification,
			&s.HumanBehaviorChanged); err != nil {
			return nil, err
		}
		samples = append(samples, s)
	}
	return samples, rows.Err()
}

// ListBehaviorDiffs returns every behavior-diff (labeled and not) with its
// before/after outputs and current labels, oldest first — the stable list the
// dashboard labeling view renders. onlyUnlabeled restricts to rows still
// needing a human label.
func ListBehaviorDiffs(ctx context.Context, pool *pgxpool.Pool, onlyUnlabeled bool) ([]DiffSample, error) {
	where := ""
	if onlyUnlabeled {
		where = "WHERE bd.human_behavior_changed IS NULL"
	}
	rows, err := pool.Query(ctx, `
		SELECT bd.id, ec.input, ec.expected_behavior,
		       before_run.actual_output, after_run.actual_output,
		       bd.judge_behavior_changed, bd.judge_justification, bd.human_behavior_changed
		FROM behavior_diffs bd
		JOIN eval_cases ec ON ec.id = bd.eval_case_id
		JOIN eval_runs before_run
		     ON before_run.prompt_version_id = bd.prompt_version_before_id
		    AND before_run.eval_case_id = bd.eval_case_id
		JOIN eval_runs after_run
		     ON after_run.prompt_version_id = bd.prompt_version_after_id
		    AND after_run.eval_case_id = bd.eval_case_id
		`+where+`
		ORDER BY bd.created_at ASC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	samples := []DiffSample{}
	for rows.Next() {
		var s DiffSample
		if err := rows.Scan(&s.BehaviorDiffID, &s.Input, &s.ExpectedBehavior,
			&s.BeforeOutput, &s.AfterOutput, &s.JudgeBehaviorChanged, &s.JudgeJustification,
			&s.HumanBehaviorChanged); err != nil {
			return nil, err
		}
		samples = append(samples, s)
	}
	return samples, rows.Err()
}

// SetHumanLabel records the INDEPENDENT human judgment of whether behavior
// changed — the gold ground-truth label, not agreement with the judge.
func SetHumanLabel(ctx context.Context, pool *pgxpool.Pool, behaviorDiffID uuid.UUID, changed bool) error {
	_, err := pool.Exec(ctx,
		`UPDATE behavior_diffs SET human_behavior_changed = $1 WHERE id = $2`, changed, behaviorDiffID)
	return err
}

// JudgeAudit is the judge's performance measured against hand labels, treating
// the human label as truth: a confusion matrix plus derived rates over the
// rows a human has labeled. This is how we decide whether the judge is
// trustworthy enough to stand in for hand-labeling at scale.
type JudgeAudit struct {
	Labeled   int     `json:"labeled"`   // rows with both a human and judge verdict
	TP        int     `json:"tp"`        // judge changed, human changed
	FP        int     `json:"fp"`        // judge changed, human not
	FN        int     `json:"fn_"`       // judge not, human changed
	TN        int     `json:"tn"`        // judge not, human not
	Agreement float64 `json:"agreement"` // (tp+tn)/labeled
	Precision float64 `json:"precision"` // of judge "changed", how many humans agreed
	Recall    float64 `json:"recall"`    // of human "changed", how many judge caught
}

// ComputeJudgeAudit tabulates the judge vs. human confusion matrix over labeled
// behavior-diffs.
func ComputeJudgeAudit(ctx context.Context, pool *pgxpool.Pool) (*JudgeAudit, error) {
	rows, err := pool.Query(ctx, `
		SELECT judge_behavior_changed, human_behavior_changed
		FROM behavior_diffs
		WHERE human_behavior_changed IS NOT NULL
		  AND judge_behavior_changed IS NOT NULL`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	audit := &JudgeAudit{}
	for rows.Next() {
		var judge, human bool
		if err := rows.Scan(&judge, &human); err != nil {
			return nil, err
		}
		audit.Labeled++
		switch {
		case judge && human:
			audit.TP++
		case judge && !human:
			audit.FP++
		case !judge && human:
			audit.FN++
		default:
			audit.TN++
		}
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}

	if audit.Labeled > 0 {
		audit.Agreement = float64(audit.TP+audit.TN) / float64(audit.Labeled)
	}
	if audit.TP+audit.FP > 0 {
		audit.Precision = float64(audit.TP) / float64(audit.TP+audit.FP)
	}
	if audit.TP+audit.FN > 0 {
		audit.Recall = float64(audit.TP) / float64(audit.TP+audit.FN)
	}
	return audit, nil
}
